### PR TITLE
Add migration for groups with invalid values

### PR DESCRIPTION
**Pull request by dhaus67**
## Description

Within the initial PR to migrate the groups bucket from a composite key created by the `storage.GroupProperties` set on the group to a UUID, groups that had an empty role name have been missed during the migration.

Those have surfaced recently after looking at a ticket ([ROX-13446](https://issues.redhat.com/browse/ROX-13446)), where the groups bucket contained two groups with properties set to `nil` and an empty role name.

Irrespective of how the groups got there in the first place, the first migration missed to add an ID to groups with a) either no role name set or b) no properties being set.

After adding validation, groups that either do not reference a role name or an auth provider are considered as invalid.

This PR creates a migration that specifically removes those groups.

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
~~- [ ] Evaluated and added CHANGELOG entry if required~~
~~- [ ] Determined and documented upgrade steps~~
~~- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into 
[rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~~

If any of these don't apply, please comment below.

## Testing Performed

- see unit tests added for the migration.

Additionally, the following tests were done manually to ensure the migration worked as expected:
1. Create a central which inserts invalid data on startup (I used a custom build that added those improper groups with nil properties and empty role name).
2. You should observe the following output when querying the groups endpoint:
```
roxcurl  /v1/groups
{"groups":[{"props":null,"roleName":""},{"props":{"authProviderId":"177332fb-c6ae-4062-ab60-54b42cdeb024","key":"","value":""},"roleName":""}
```
3. Upgrade to this version.
4. Query the groups endpoint and verify that the previously invalid entries are not available anymore.
```
roxcurl  /v1/groups
{}
```

### CODE REVIEW

**Comment by dhaus67** [comment]
Current dependencies on/for this PR:
* master
  * **PR #3789** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/3789" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/stackrox/stackrox/3789?utm_source=stack-comment).

File: migrator/migrations/m_111_to_m_112_groups_invalid_values/migration.go
Comment (ivan-degtiarenko) [inline migrator/migrations/m_111_to_m_112_groups_invalid_values/migration.go:61]: Should we also delete groups with empty role names here as a pre-caution?

File: migrator/migrations/m_111_to_m_112_groups_invalid_values/migration.go
Comment (ivan-degtiarenko) [inline migrator/migrations/m_111_to_m_112_groups_invalid_values/migration.go:91]: There's no `2`

File: migrator/migrations/m_111_to_m_112_groups_invalid_values/migration.go
Comment (ivan-degtiarenko) [inline migrator/migrations/m_111_to_m_112_groups_invalid_values/migration.go:77]: let's explicitly return `nil` instead of `err` here

File: migrator/migrations/m_111_to_m_112_groups_invalid_values/migration.go
Comment (ivan-degtiarenko) [inline migrator/migrations/m_111_to_m_112_groups_invalid_values/migration.go:88]: What if we use `for _, group := range groupStoredByCompositeKeys`? Then we don't need to address array value by index within the loop

File: migrator/migrations/m_111_to_m_112_groups_invalid_values/migration.go
Comment (dhaus67) [reply migrator/migrations/m_111_to_m_112_groups_invalid_values/migration.go:61]: We do delete groups with empty role names as long as they are still stored keyed by the composite key.

I don't think we need to remove groups with empty role names when they are stored using a unique ID - currently it would be possible to delete those via API directly (I also don't expect those to be there, but then again neither did I with the current situation). Still, I'd say let's not do this and add additional complexity to the migration.

Comment (ivan-degtiarenko) [reply migrator/migrations/m_111_to_m_112_groups_invalid_values/migration.go:62]: For me, empty role name = invalid group = should be deleted. This corresponds with the goal of this migration as I see it - fix our previous mistake and return the groups bucket to the valid state. Maybe I understood it wrong and we delete somewhere groups with an empty role, but definitely not during startup:
https://github.com/stackrox/stackrox/blob/6113b92dce180cd7856b1917569fb3fd18786037/central/group/datastore/singleton.go#L22-L27

I won't fixate on this though. If you feel strongly about not adding this to migration right now, let's put it on hold - for me, it makes sense to do it now due to our release cadence.

File: migrator/migrations/m_111_to_m_112_groups_invalid_values/migration_test.go
Comment (ivan-degtiarenko) [inline migrator/migrations/m_111_to_m_112_groups_invalid_values/migration_test.go:91]: super-nit

```suggestion
	// 1.  Migration should succeed if the bucket does not exist
```

File: migrator/migrations/m_111_to_m_112_groups_invalid_values/migration.go
Comment (ivan-degtiarenko) [inline migrator/migrations/m_111_to_m_112_groups_invalid_values/migration.go:94]: nit: can be inlined

**Review by ivan-degtiarenko — APPROVED** [review]
LGTM modulo couple of nits

File: migrator/migrations/m_111_to_m_112_groups_invalid_values/binenc/byte_slice_list.go
Comment (janisz) [inline migrator/migrations/m_111_to_m_112_groups_invalid_values/binenc/byte_slice_list.go:31]: Who is using it?

File: migrator/migrations/m_111_to_m_112_groups_invalid_values/binenc/byte_slice_list.go
Comment (janisz) [inline migrator/migrations/m_111_to_m_112_groups_invalid_values/binenc/byte_slice_list.go:39]: Who is using it?

File: migrator/migrations/m_111_to_m_112_groups_invalid_values/binenc/byte_slice_list.go
Comment (dhaus67) [reply migrator/migrations/m_111_to_m_112_groups_invalid_values/binenc/byte_slice_list.go:31]: I followed the same examples for copying stackrox code into the migrations - which is copying the package as is and not modifying it, even though some functions may be unused.

File: migrator/migrations/m_111_to_m_112_groups_invalid_values/binenc/byte_slice_list.go
Comment (dhaus67) [reply migrator/migrations/m_111_to_m_112_groups_invalid_values/binenc/byte_slice_list.go:39]: Same as above.

**Pull request merged**
